impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java

```diff
@@ -7,39 +7,88 @@
  * The caller is guaranteed that all allocated port numbers are unique, even across invocations of #preAllocateListeningSockets.
- *
+ * <p>
  * Before the caller attempts to bind, they must call #close.  Once close is called no further use of the socket
  * allocator is permitted.
  */
 public class ListeningSocketPreallocator implements AutoCloseable {
 
+    private static final int PORT_RANGE_LOW = 10_000;
+    // Ideally this would be derived from `sysctl net.inet.ip.portrange.first` or `sysctl net.ipv4.ip_local_port_range` depending on the platform
+    // however that's not so easily achieved, so we hard code instead.
+    private static final int PORT_RANGE_HIGH = 30_000;
+    private static final int MAX_PORT_COUNT = PORT_RANGE_HIGH - PORT_RANGE_LOW;
+
     private final List<ServerSocket> all = new ArrayList<>();
 
+    private final Random random = new Random();
+
     /**
      * Instantiates a new Listening socket preallocator.
      */
     public ListeningSocketPreallocator() {
     }
 
+    /**
+     * Pre-allocate 1 or more ports from a defined range, to avoid collisions without going network connections,
+     * which are available for use once the #close method is called.
+     *
+     * @param num number of ports to pre-allocate
+     * @return a List of bound sockets
+     */
+    public List<ServerSocket> preAllocateListeningSockets(int num) {
+        if (num < 1) {
+            return List.of();
+        }
+        if (all.size() + num > MAX_PORT_COUNT) {
+            throw new IllegalArgumentException("Can't request more than " + MAX_PORT_COUNT + " ports. There are already " + all.size() + " allocated.");
+        }
+        return random.ints(PORT_RANGE_LOW, PORT_RANGE_HIGH)
+                .mapToObj(number -> {
+                    try {
+                        var serverSocket = new ServerSocket(number);
+                        serverSocket.setReuseAddress(true);
+                        if (serverSocket.isBound()) {
+                            return serverSocket;
+                        }
+                    }
+                    catch (IOException e) {
+                        if (e instanceof BindException) {
+                            return null;
+                        }
+                        else {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                    return null;
+                }).filter(Objects::nonNull)
+                .peek(all::add)
+                .limit(num)
+                .collect(Collectors.toList());
+    }
+
     /**
      * Pre-allocate 1 or more ephemeral ports which are available for use once the #close method is called.
      *
      * @param num number of ports to pre-allocate
      * @return stream of ephemeral ports
      */
-    public Stream<ServerSocket> preAllocateListeningSockets(int num) {
+    public List<ServerSocket> preAllocateEphemeralListeningSockets(int num) {
```

Comment:
We don't need two competing implementations however I've kept this here for now while we settle on one.